### DISCUSSION TITLE
Identifying peptide oligomeric states using Python LOOS

### DISCUSSION THREAD

**Discussion by acnash** [GrossfieldLab/loos#51] [Users]
Dear LOOS developers and users,

I thought I would put up a discussion here to avoid reinventing the wheel (in case it exists). 

**Objective:**
With the help of LOOS I'm looking at writing a python function that will report the oligomeric state of a system. 

**System:**
Using Gromacs, I'm running a martini CG simulation of four transmembrane (TM) helices in a POPC:DPPC bilayer system. The simulation runs for N steps and saves M frames to a .trr gromacs trajectory. 

**Outcome:**
On a single coordinate frame by coordinate frame basis (in the .trr file) I would like to report the oligomeric state of the four TM helices. Thus, by the end of the trajectory, I should have M (number of frames) rows of oligomeric state data. For examples, a single saved coordinate frame will of four helices report one of four oligomeric states:

- 1:1:1:1 (all are monomeric)
- 1:3 (monomer and a trimer)
- 2:2 (two dimers)
- 4 (all four are together)  

I plan on automating these processes to run over several hundred short CG MD simulations. I'm happy doing the python grunt work, I would like to know (a) doing this exist in any shape or form already, (b) if not, which particular LOOS classes/functions would you envisage using?

I imagine an interaction distance of e.g., 1.4 nm, between n helices is essential. If within or less than that distance, those n helices are considered to be in one of the four above states. 

Many thanks
Anthony Nash
University of Oxford

**Comment by agrossfield** [comment]
HI Anthony,

There isn't a tool that does exactly what you want, but the pieces are there to make it easy. Almost everything you need is in the class `AtomicGroup`. The steps as I'd see them:
- Use gmxdump2pdb.pl to make a fake psf to use as the system file instead of the gro file. This has the advantage of having connectivity information as well as a segid (which is useful for selection.
- Read the system in.
- Select the protein with something like segid
```python
all_chains = loos.selectAtoms(system, 'segid ="PRTN"')
chains = all_chains.splitByMolecule() # or, if you're using a gro file, you'll need to do something else to get the individual chains
```
- As far as the actual logic, I see 2 possibilities. The first would just be the centroid-centroid distance between them, which would look something like
```python
c0 = chains[0].centroid()
c1 = chains[1].centroid()
dist_sq = c0.distance2(c1, system.periodicBox()) # compute the distance squared, respecting periodicity
if (dist_sq < threshold_sq): 
    # do something to store that they're in contact

```
This will work fine if you know that the helices are going to stay more or less vertical in the membrane. If you don't want to assume that, then you could use `AtomicGroup.contactWith()`, which does the double loop over atoms in 2 groups, and returns true if at least a certain number of atoms pairs are within a given distance. This approach will be a bit slower, but it'll handle arbitrary configurations more gracefully. Again, you'll want to explicitly feed it the periodic box information -- almost all distance calculations in loos have 2 versions, one that computes the simple distance and one that respects periodicity. 

I can see a couple of ways to figure out how many aggregates there are given the identification of pairs. I'll lay them out if you'd like, but I suspect you already have similar or better ideas. There might even be something in scipy (clustering, perhaps?) that does it, but with only 4 chains I suspect that would be overkill.

Just as an aside, I'd recommend using xtc instead of trr for your trajectories. If I recall correctly, trr is a much more verbose file format, and your storage will quickly become an issue. 

Hope this helps (and please ask more questions if it wasn't!),

Alan

PS I could easily imagine this being generally useful. Let me know if you're interested in adding the tool to LOOS when it's done.

**Reply by acnash** [reply]
Hi Alan, that's great. Thank you for all this information, it's very helpful. 

I'm more than happy to add this tool into LOOS after I've finished refining it. Before then, a quick question if you don't mind. In your example code above you have the line:

`chains = all_chains.splitByMolecule()`

I am able to generate chain IDs (i.e., 4 proteins would give me chains A, B, C and D), using gmx trjconv -f ...gro -s ...tpr, however, I can't figure out if there is a mechanism I'm missing that will allow the very same chain IDs when I use:

gmx dump -s foo.tpr | gmxdump2pdb.pl test

It seems that only the trjconv tool can add chain identifiers. To be honest, this isn't a big problem as I can just group by atoms, but for the sake of completeness, I was wondering whether you know how to add in the chain IDs?

Thanks

**Update:** to get chain IDs into the pdb I have used the output pdb from gmxdump2pdb and passed it into trjconv whilst pointing back at the original tpr file and specifying the CONECT option. I've yet to now see whether this new pdb file is compatible with the psf file. The steps were:

`gmx dump -f npt_prod.tpr | gmxdump2pdb.pl test` #gives me a test.pdb and test.psf
`gmx trjconv -f test.pdb -s npt_prod.tpr -CONECT -o test2.pdb`

Now I have A, B, C and D chain labels representing my four helices.

**Reply by agrossfield** [reply]
Good solution. We have a tendency to use SEGID instead of the CHAINID. While you can select on chainid in loos, it's restricted to 2 columns (and IIRC vmd only reads 1 column), while segid is 4 characters. This is one of the reasons we like the psfs. Take a look at `gmxdump2pdb.pl -h`, particularly the `--segmap` option, which gives you a way to control the segid.

**Reply by agrossfield** [reply]
BTW, `splitByMolecule` works on connectivity -- it recurses through the list of bonds -- not chainid. So, it ought to work even if the chainid or segid isn't set.

**Reply by acnash** [reply]
Sorry if this is premature, but I'm seeing unexpected connectivity after running gmxdump2pdb.pl (with no options applied). I'm working on the assumption that the output from gmx dump mytpr.tpr was fine and the connectivity in my simulations work. I have no reason to doubt this.

I noticed that `splitByMolecule` wasn't returning a list of length 4, one for each TM helices as I would expect. Rather, the returning list contained each of the amino acids instead. I looked at my PDB file (this is separate from my tinkerings with `chainid` in my previous post) and noticed that the backbone bond between amino acids in my peptides was not in the connectivity segment of the PDB file. 


ATOM   **9506**   BB  LYS 6932      21.730  14.730  46.733  1.00  1.00      PRTA
ATOM   9507  SC1  LYS 6932      24.778  13.575  47.188  1.00  1.00      PRTA
ATOM   9508  SC2  LYS 6932      26.643  11.480  47.111  1.00  1.00      PRTA
ATOM   **9509**   BB  ARG 6933      19.610  13.320  48.502  1.00  1.00      PRTA
ATOM   9510  SC1  ARG 6933      16.565  13.032  48.832  1.00  1.00      PRTA
ATOM   9511  SC2  ARG 6933      13.768  11.057  48.027  1.00  1.00      PRTA
....
CONECT **9506** 9507
CONECT 9507 9508
**_<I believe there should be an entry here...?>_**
CONECT **9509** 9510
CONECT 9510 9511


As you can see, this is a CG model. I've double-checked the output from gmxdump2pdb and I see the expected output (see below). Notice Protein_A is reporting 4 individual molecules even though the BB-BB bond is not present in the output:


Found segment 'DPPC' [type=0] (0)
        Molecules: 132
Found segment 'POPC' [type=1] (1)
        Molecules: 102
Found segment 'WN' [type=2] (2)
        Molecules: 6465
Found segment 'NA' [type=3] (3)
        Molecules: 106
Found segment 'CL' [type=4] (4)
        Molecules: 126
Found segment 'Protein_A' [type=5] (5)
        **Molecules: 4**
Found 6 segments in topology
Found 6 segments defined

I then did some digging and found that the BB-BB bond in my gromacs topology file is handled with constraints, on top of an elastic bond network for everything else. So I added -constraints option to the gmxdump2pdb.pl command but the BB-BB connectivity is still not present, so splitting by molecule after filtering on segid returns a list of amino acids rather than a list of peptides.

For my task, I don't think bond connectivity is going to make up a big part. The four TM helices are all identical, so it's simply a list of the CG peptide beads divided by 4. 

I just wanted to flag and I'm sorry in advance if I missed something obvious. Thanks!

**Reply by agrossfield** [reply]
I'm glad you have a workaround, but of course we want the code the to do the right thing for everyone.

Possibly stupid suggestion: the option should have 2 dashes (`--constraints`) not one, the way you typed it in the previous note. 

Possible dumb suggestion/question number 2: does the same thing happen if you use the psf as your system file? You could use the `--segmap` option to set segid's to mimic what you were using the chainids for.

If the problem persists, could you send us the tpr file and gromacs version, so we can try to figure out what's going wrong for you?

**Reply by acnash** [reply]
Thank you for your help. You were right, there was a "-" typo. However, when I corrected the command it still failed to output the expected bond information. 

For reproducibility please see below. Gromacs version 2021.2, the latest build released near the beginning of May. The .tpr and the resulting .pdb and .psf I've managed to create are stored on my Google Drive:
[Google Drive Link](https://drive.google.com/drive/folders/1a97J2utFcSTh5_DGl-nA-sIto3lUO8qO?usp=sharing)

**gmxdump2pdb.pl command:**

`gmx dump -s ../EPOC_0/0/npt_prod.tpr | ~/miniconda3/envs/soft-machine/bin/gmxdump2pdb.pl --constraints test`

**python code:**
```
#There are 26 amino acids per helix i.e., 104 amino acids/resides

import loos
import loos.pyloos
import sys

#create system and load in trajectory
system = loos.createSystem("test.psf")
#traj = loos.pyloos.Trajectory("../EPOC_0/0/npt_prod.trr", system)


#select "domains"
allChainsAG = loos.selectAtoms(system, 'segid=="PRTA"')  #contains all 236 beads
chainsList = allChainsAG.splitByMolecule() #this splits into 136 parts

#the first part (chainsList[0]) is a <GROUP> that represents the three beads in the first residues (LYS)
print(chainsList[0])

#this should return 4 (one per helix), not 136 (one per residue, even those residues might be broken as each helix is only 26 residues
#long, which chould be 104, not 136) - I think this is caused by no bond info between BB-BB beads and some other beads (perhaps)..
print(len(chainsList))
```

**Update:** if it helps, further information. The system is a DPPC, POPC and Water (with charge) model with 4 identical alpha-helices. The system was encoded using the latest version of Martini (a release this year). In this instance, the four helices (all identical) are: KRSNFLALLLGIYLGLSLVFKNSRK

**Reply by agrossfield** [reply]
Thank you for the super-clear bug report! I'll take a look, but chances are this will need @tromo (Tod), the original author of the script.

**Reply by acnash** [reply]
You're welcome. I hope it helps. Disclaimer: there is the possibility that I've done something wrong, but I'm always happy to be corrected.  In the meantime, I can work by residue count and divide the system up accordingly.

**Reply by acnash** [reply]
Final update. I've looked at pulling out the individual AtomicGroups that represent the BB bead for each of the four helices. As each helix is identical in length (26), I thought a for loop through the first 26 residue atomic groups (based only on the bond connectivity) would prove to be a simple workaround. It turns out that I'm looking at 34 atomic groups rather than the expected 26. Those additional atomic groups look to be empty:

```
<GROUP PERIODIC='(101.969,78.428,137.396)'>
   <ATOM INDEX='9531' ID='9532' NAME='BB' RESID='6944' RESNAME='TYR' COORDS='(20.94,12.994,66.617)' VELOCITIES='(0,0,0)' ALTLOC='' CHAINID='' ICODE='' SEGID='PRTA' B='1' Q='1' CHARGE='0' MASS='1' ATOMICNUMBER='-1' MASK='203'>
  <BOND>9533</BOND>
</ATOM>
</GROUP>
<GROUP PERIODIC='(101.969,78.428,137.396)'>
</GROUP>
<GROUP PERIODIC='(101.969,78.428,137.396)'>
</GROUP>
<GROUP PERIODIC='(101.969,78.428,137.396)'>
</GROUP>
<GROUP PERIODIC='(101.969,78.428,137.396)'>
   <ATOM INDEX='9536' ID='9537' NAME='BB' RESID='6945' RESNAME='LEU' COORDS='(23.037,11.067,67.843)' VELOCITIES='(0,0,0)' ALTLOC='' CHAINID='' ICODE='' SEGID='PRTA' B='1' Q='1' CHARGE='0' MASS='1' ATOMICNUMBER='-1' MASK='203'>
  <BOND>9538</BOND>
</ATOM>
</GROUP>
```

Using the files I linked above, the test code for this:

```
#create system and load in trajectory
systemAG = loos.createSystem("test.pdb")
#traj = loos.pyloos.Trajectory("../EPOC_0/0/npt_prod.trr", system)

#select "domains"
allChainsAG = loos.selectAtoms(systemAG, 'segid=="PRTA"')  #contains all 236 beads
chainsList = allChainsAG.splitByMolecule() #this splits into 136 parts (it should be 4*26 = 104)

countInt: int = 0
for residueAG in chainsList:
    bbAG = loos.selectAtoms(residueAG, 'name=="BB"')
    print(bbAG)
    countInt = countInt + 1
    if countInt > 26:  #this should be ok, but the additional empty AtomicGroups means 26 is too short
        break
```

Again, this may be an unexpected consequence of something I've done wrong from the start, but I thought further information can only help.

**Reply by agrossfield** [reply]
I think your problem might be related to issue #47, which we just fixed. In any event, @tromo checked and the current version works correctly, while he can reproduce the error if he rolls it back. 

So, I think all you need to do is pull and (optionally) reinstall:
```
git pull
scons PREFIX=$CONDA_PREFIX install
```

(you will pick up some new stuff, so the build will take a minute)

Let me know what happens -- if it works, I'll mark this as solved.

**Reply by acnash** [reply]
That's great. I'll give this a try and let you know how it goes.

**Reply by acnash** [reply]
Just an update as I understand I've left this issue hanging. I've just finished building a new GPU server. Hopefully, over the next week, I'll find the time to try out this change.
